src/bonsai/bonsai/bim/module/sequence/operator.py

```diff
@@ -654,6 +654,75 @@ def execute(self, context):
         return {"FINISHED"}
 
 
+class ImportAlignmentCSV(bpy.types.Operator, tool.Ifc.Operator, ImportHelper):
+    bl_idname = "bim.import_alignment_csv"
+    bl_label = "Import Alignment CSV"
+    bl_options = {"REGISTER", "UNDO"}
+    filename_ext = ".csv"
+    filter_glob: bpy.props.StringProperty(default="*.csv", options={"HIDDEN"})
+
+    @classmethod
+    def poll(cls, context):
+        ifc_file = tool.Ifc.get()
+        if ifc_file is None:
+            cls.poll_message_set("No IFC file is loaded.")
+            return False
+        return True
+
+    def _execute(self, context):
+        import ifcopenshell.api.alignment
+        self.file = tool.Ifc.get()
+        start = time.time()
+        alignment = ifcopenshell.api.alignment.create_alignment_from_csv(self.file,self.filepath)
+
+        # IFC 4.1.5.1 alignments cannot be contained in spatial structures, but can be referenced into them
+        sites = self.file.by_type("IfcSite")
+        for site in sites:
+            ifcopenshell.api.spatial.reference_structure(self.file,products=[alignment],relating_structure=site)
+
+        # process the generated IfcReferent for the alignment
+        for rel in alignment.IsNestedBy:
+            for referent in rel.RelatedObjects:
+                if referent.is_a().upper() == "IFCREFERENT":
+                    referent_obj = bpy.data.objects.new(tool.Loader.get_name(referent),None)
+                    tool.Geometry.link(referent,referent_obj)
+                    tool.Collector.assign(referent_obj,should_clean_users_collection=False)
+
+        # an alignment can be an aggregation of multiple child alignments (ie. multiple verticals for a single horizontal)
+        curves = []
+        for rel in alignment.IsDecomposedBy:
+            for agg in rel.RelatedObjects:
+                if agg.is_a().upper() == "IFCALIGNMENT":
+                    curves.append(agg.Representation.Representations[0].Items[0]) # 3D curve
```

Comment:
I find this risky, it assumes 1) there is a rep 2) the 0th rep is the appropriate context and 3) the 0th item is a 3d curve. Better to use `util.representation.get_representation`. Don't know enough whether or not the curve assumption is safe.

Reply:
Refactored so its not so risky.